### PR TITLE
automatically create dynamic tables

### LINKED ISSUE #234: get_class for DynamicTable
Right now `get_class` is not working for extensions that inherit from the `DynamicTable` class. Writing an extension is pretty involved, requiring you to register a class and a mapper, but the workflow is pretty uniform and we should be able to automate it.

Briefly describe the needed feature as well as the reasoning behind it

**Is your feature request related to a problem? Please describe.**
A clear and concise description of what the problem is. Ex. I'm trying to achieve [...]

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

## Checklist

- [ ] Have you ensured the feature or change was not already [reported](https://github.com/hdmf-dev/hdmf/issues) ?
- [ ] Have you included a brief and descriptive title?
- [ ] Have you included a clear description of the problem you are trying to solve?
- [ ] Have you included a minimal code snippet that reproduces the issue you are encountering?
- [ ] Have you checked our [Contributing](https://github.com/hdmf-dev/hdmf/blob/dev/docs/CONTRIBUTING.rst) document?

### PR DESCRIPTION
**Pull request by bendichter**
## Motivation

fix #234

be able to use `get_class` on dynamic tables.

Works and tested for:
* regular columns
* DynamicTableRegion columns
* indexed columns

### CODE REVIEW

**Comment by rly** [comment]
This is tricky because it hardcodes DynamicTable into the BuildManager/TypeMap and also the `__get_cls_dict` function is already difficult to maintain, with its handling of normal Container fields config, MultiContainerInterface config, and DynamicTable column config. Let me think about it, try something, and get back to you. It might not be worth extracting into something like the modular ObjectMapper, given that there are only three config types at the moment, but it might be straightforward and worth the modularity and cleanliness.

**Comment by bendichter** [comment]
@rly I agree it is becoming a lot. Right now I am modularizing more horizontally. For instance, I have factored out the creation of docval into its own method. But here you still need a bunch of if statements to deal with multi container interfaces and dynamic tables.

**Comment by bendichter** [comment]
@rly this is working now. Could you take a look? ndx-photometry depends on this.

fyi @akshay-jaggi

File: src/hdmf/build/manager.py [hunk 1]
```diff
@@ -574,6 +574,10 @@ def _build_docval(self, base, addl_fields, name=None, default_name=None):
         """
         docval_args = list(deepcopy(get_docval(base.__init__)))
         for f, field_spec in addl_fields.items():
+            if base.__name__ == 'DynamicTable' \
+                    and getattr(field_spec, 'neurodata_type_inc', None) in (
+                    'VectorData', 'DynamicTableRegion'):
+                continue  # do not add DynamicTable columns to init docval
```
Comment (ajtritt) [inline src/hdmf/build/manager.py:580]: Why isn't this getting hit?

**Comment by ajtritt** [comment]
> ndx-photometry depends on this.

I would like to understand why an extension depends on an autogenerated class...

**Comment by bendichter** [comment]
@ajtritt 
> Why isn't this getting hit?

Good catch! https://github.com/hdmf-dev/hdmf/pull/522/commits/a2c44712522b1ee48447371d0e3729ae5566bc28 fixes it.

> I would like to understand why an extension depends on an autogenerated class...

The current implementation does not define custom API classes for each DynamicTable. We could of course define custom APIs if we need to, but I'd rather auto-generate if there isn't anything special going on.

**Comment by bendichter** [comment]
@rly if I resolve these conflicts, would you be OK with merging this? We can change it to your refactor once that is ready.

**Comment by yarikoptic** [comment]
FWIW -- I was told that this PR should (didn't test) resolve inability of dandi-cli to open the nwb file with `ndx_simulation_output`, so that basic metadata (for `dandi organize`) could be read.

<details>
<summary>ATM we would crash on `NWBHDF5IO(path, "r", load_namespaces=True)` </summary> 

```shell
  File "/home/yoh/proj/dandi/dandi-cli-master/dandi/pynwb_utils.py", line 182, in _get_pynwb_metadata
    with NWBHDF5IO(path, "r", load_namespaces=True) as io:
  File "/usr/lib/python3/dist-packages/hdmf/utils.py", line 580, in func_call
    return func(args[0], **pargs)
  File "/home/yoh/proj/dandi/dandi-cli-master/venvs/dev3/lib/python3.9/site-packages/pynwb/__init__.py", line 230, in __init__
    super(NWBHDF5IO, self).load_namespaces(tm, path, file=file_obj)
  File "/usr/lib/python3/dist-packages/hdmf/utils.py", line 580, in func_call
    return func(args[0], **pargs)
  File "/usr/lib/python3/dist-packages/hdmf/backends/hdf5/h5tools.py", line 120, in load_namespaces
    return cls.__load_namespaces(namespace_catalog, namespaces, f)
  File "/usr/lib/python3/dist-packages/hdmf/backends/hdf5/h5tools.py", line 168, in __load_namespaces
    d.update(namespace_catalog.load_namespaces('namespace', reader=reader))
  File "/usr/lib/python3/dist-packages/hdmf/utils.py", line 580, in func_call
    return func(args[0], **pargs)
  File "/usr/lib/python3/dist-packages/hdmf/build/manager.py", line 460, in load_namespaces
    deps = call_docval_func(self.__ns_catalog.load_namespaces, kwargs)
  File "/usr/lib/python3/dist-packages/hdmf/utils.py", line 423, in call_docval_func
    return func(*fargs, **fkwargs)
  File "/usr/lib/python3/dist-packages/hdmf/utils.py", line 580, in func_call
    return func(args[0], **pargs)
  File "/usr/lib/python3/dist-packages/hdmf/spec/namespace.py", line 490, in load_namespaces
    ret[ns['name']] = self.__load_namespace(ns, reader, types_key, resolve=resolve)
  File "/usr/lib/python3/dist-packages/hdmf/spec/namespace.py", line 447, in __load_namespace
    catalog.register_spec(spec, spec_file)
  File "/usr/lib/python3/dist-packages/hdmf/utils.py", line 580, in func_call
    return func(args[0], **pargs)
  File "/usr/lib/python3/dist-packages/hdmf/spec/catalog.py", line 45, in register_spec
    raise ValueError("'%s' - cannot overwrite existing specification" % type_name)
ValueError: 'Data' - cannot overwrite existing specification

> /usr/lib/python3/dist-packages/hdmf/spec/catalog.py(45)register_spec()
-> raise ValueError("'%s' - cannot overwrite existing specification" % type_name)

```
</details>

 I would appreciate if the verdict on this PR was made or some alternative fix implemented so that hdmf could load the file, and these data could be organized and uploaded to the archive.

**Comment by rly** [comment]
This has been superseded by #536.

**Pull request closed without merging**